### ISSUE TITLE
InstantRun fails with class signature mismatch

### ISSUE THREAD

**Issue by bduhbya** [AzureAD/azure-activedirectory-library-for-android#724]
When trying use Instant run on Android, the build will fail with:

Error:Execution failed for task ':app:transformClassesWithInstantRunForDevDebug'.

> class "com.microsoft.aad.adal.AuthenticationCallback"'s signer information does not match signer information of other classes in the same package

I verified that removing the security signatures from the class.jar manifest allows the build to succeed. 

This may seem like a trivial issue; however, instant run has the potential to save significant developer hours.

Thank you.

**Comment by zedixns** [comment]
It was already noted here: #720

**Comment by bduhbya** [comment]
Thanks, Vladimir. It came up in the middle of some other tasks and another
team member asked me to file it. My appologies for the duplicate.

**Comment by weijjia** [comment]
class.jar is a signed jar. Can you paste the full error info for the task app:transformClassesWithInstantRunForDevDebug? Also, can you paste the app dependency tree here?

**Comment by bduhbya** [comment]
That is the extent of the error. It is saying that it thinks at least one of the .class files in the classes.jar file is not signed with the same cert specified by the manifest (in my understanding).

When I get into the office I can give you a stack trace into the Android code that throws the error.  If this is an Android tools bug, then the trace may be useful.  Otherwise I don't know if it will give you any insights but to show you how the do the comparison. 

When you ask for the dependency tree, you want all third party libraries consumes by the app?  I can see about providing it for two projects that experience this issue.  As mentioned previously, removing the manifest signature manually does allow the instant run build task to succeed so there could be an issue with the signing process.  Are you thinking there could be confusion in the tool with other class files from other jar/aar libs?

If you build a simple project with adal and the most up to date Android studio and gradle/gradle plugins are you able to reproduce the failure by enabling instant run and pressing play to deploy the test app?

To recap, when I'm back in the office, I will see about providing a list of each app's consumed libraries and I will give you the verbose trace into the Android tooling java sources for the error.

Thanks

Regards,

Brent

**Comment by zedixns** [comment]
@weijjia 
I was getting the same failure as Brent there with a class not being able to verify,but now I am getting a different when trying to use instant run.
Dono if this helps but here you go:

Error:Execution failed for task ':app:transformClassesWithInstantRunForDistributionDebug'.
> class "com.microsoft.aad.adal.R$dimen"'s signer information does not match signer information of other classes in the same package

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:transformClassesWithInstantRunForDistributionDebug'.
> class "com.microsoft.aad.adal.R$dimen"'s signer information does not match signer information of other classes in the same package

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

**Comment by bduhbya** [comment]
My Android studio has updated twice since I first saw the error. The --stacktrace option does not seem to be working as it did previously.  I will privately send you the deps.

**Comment by DevilGeek** [comment]
I have the same problem

`class "com.microsoft.aad.adal.R$dimen"'s signer information does not match signer information of other classes in the same package`

Does anyone have a solution ?

**Comment by imakhalova** [comment]
+1 who has repro, I'd like it to be fixed as disabling Instant Run is not a good solution

**Comment by addev-ashish** [comment]
@bduhbya: did you get a chance to send across the deps?

**Comment by gauravgaikwad777** [comment]
+1 
![253fa7e8-8967-11e6-9f4b-4305fda49da2](https://cloud.githubusercontent.com/assets/8176450/19508733/a601f0a0-958f-11e6-8321-bbbb132a3cb9.png)
Because of this I went back to Android Studio 2.1.3, and now with MacOS Sierra update even Android Studio 2.1.3 is crashing always and they only fix crashes for latest release which is 2.2.
To upgrade to 2.2 I need this fix.

**Comment by bduhbya** [comment]
I did not. Been working two projects and it fell off my radar.  Let me send them now.

**Comment by bduhbya** [comment]
@addev-ashish, I just sent the deps to the ADAL alias.  Let me know if you need more data.  I sent the basic deps listed in our gralde.

**Comment by addev-ashish** [comment]
@gauravgaikwad777 if you are absolutely blocked, please turn off Instant run from File->settings->Build, Execution, Deployment->Instant Run

Thanks @bduhbya 

I am actively working on this as the highest priority. Currently its looking like a bug in instant run's code. First InstantRun should not try to modify binaries on included packages, and even if it does it should use the appropriate class loader, currently looking at their code it looks like they are using the class loader of the main thread.

I have filed a bug on Google (providing logs) to take a look at this, we are tracking it on https://code.google.com/p/android/issues/detail?id=225547 
Will add more data to that bug as my investigation proceeds. Currently I am trying to get a simplified repro i can hand off to Google.

**Comment by addev-ashish** [comment]
@bduhbya @gauravgaikwad777 @imakhalova @zedixns 
Hi Guys,
We have been investigating this at our end, we do not have a solution yet and Google has not responded to the bug again. But there is a workaround that is working for me. I would like to validate if this works for anyone else here. Kindly try it and update back on this issue.

Steps:
1. Remove the ADAL dependencies entry from build.gradle of your APK, validate you get compile time errors
2. Go to maven central and download ADAL to a temp location (say c:\temp)
3. In your project select File->New->New Module
4. From the Dialog select "Import .JAR/.AAR Package"
5. In the next screen type the path to adal-1.10.0.aar downloaded in step 2
6. Go to File->Project Structure
7. Select your project, the one that generates the APK and in the right pane select "Dependencies"
8. Hit the Green '+' on the far right side of the screen.
9 Select ":adal:1.10.0"

Once the above is done, your code should compile like before and InstantRun should work.

@DevilGeek: your problem looks slightly different, but I would be interested in knowing if the workaround works for you.

Note: I am not closing this issue, we want to fix this right, but we want to unblock all of you as well.

**Comment by addev-ashish** [comment]
I was able to repro this with a simple JAR with ADAL out of the picture, IMO the following is the issue.
If 2 classes in different packages have the same name (in this case AuthenticationCallback) and if one of them is signed. Any attempt to use the class from the signed jar will result in InstantRun failing.
And the failure is due to Reflection failing to load the class as it is signed with a different signature than the app's classes.

**Comment by addev-ashish** [comment]
Google has accepted this is a bug with Android Studio. They expect the fix to roll out with 2.3. Here is the update from the bug in Android Studio Database.
======
change: Iddc9d100cad560da1a2aad141039b874af739c8b
First of all, thanks for the bug report. The fix should be out in 2.3.0-alpha3 or 2.3.0-beta1.
Status: FutureRelease 
=====

I am closing this issue as there is nothing else we can do here.

**Comment by imakhalova** [comment]
Hey guys, why do we sign this .class if it's not going to be verified? Can we exclude this step from process and leave only aar signature?

**Comment by striboldt** [comment]
I'm still able to reproduce the issue with 2.3.0-beta1. Anyone able to confirm the fix in 2.3.0-beta1 ?

**Comment by Tomanow** [comment]
I am still experiencing the issue with 2.3.0-beta1 as well... Re-open?

**Comment by zedixns** [comment]
**It did not make it to 2.3.0-beta1. 2.4.0-alpha1 will have it.**

Copied comment from the issue reported on: https://code.google.com/p/android/issues/detail?id=225547
